source/extensions/priv/server/ntds_decrypt.c

```diff
@@ -0,0 +1,194 @@
+/*!
+* @file ntds_decrypt.c
+* @brief Definitions for NTDS decryption functions
+*/
+#include "precomp.h"
+
+/*!
+* @brief Convert bytes into a Hex string representing those bytes.
+* @param data Pointer to the byte array we are converting
+* @param length Integer representing the length of the byte array
+* @param output Pointer to the string we are outputting the result to
+*/
+void bytes_to_string(LPBYTE data, int length, LPSTR output){
+	for (int i = 0; i < length; i++){
+		sprintf(output + (i << 1), "%02X", ((LPBYTE)data)[i]);
+	}
+}
+
+/*!
+* @brief Decrypt an  encrypted LM or NT Hash.
+* @param encryptedNTLM Pointer to an encryptedhash struct for the LM or NT hash we wish to decrypt.
+* @param pekDecrypted Pointer to a decryptedPEK structure that holds our decrypted PEK
+* @param hashString Pointer to the string where we will store the decrypted hash
+* @param rid DWORD representing the Relative ID(RID) of the account
+* @returns Indication of sucess or failure.
+*/
+BOOL decrypt_hash(struct encryptedHash *encryptedNTLM, struct decryptedPEK *pekDecrypted, char *hashString, DWORD rid){
+	BOOL cryptOK = FALSE;
+	BYTE encHashData[NULL_TERIMNATED_HASH_LENGTH] = { 0 };
+	BYTE decHash[NULL_TERIMNATED_HASH_LENGTH] = { 0 };
+
+	memcpy(&encHashData, &encryptedNTLM->encryptedHash, HASH_LENGTH_BYTES);
+	cryptOK = decrypt_rc4(pekDecrypted->pekKey, encryptedNTLM->keyMaterial, encHashData, 1, HASH_LENGTH_BYTES);
+	if (!cryptOK){
+		return FALSE;
+	}
+	cryptOK = decrypt_hash_from_rid(encHashData, &rid, decHash);
+	if (!cryptOK){
+		return FALSE;
+	}
+	bytes_to_string(decHash, HASH_LENGTH_BYTES, hashString);

```

Comment:
You're copying having into account your source length, but not checking the dst (hashString) length. I guess hashString will be long enough on your callers, just warning!